LANGUAGE_SDK_SPECIFICATION.md

```diff
@@ -0,0 +1,1325 @@
+- How to refactor non-deterministic code
+- Examples of correct and incorrect patterns
+
+The SDK SHOULD provide:
+
+- Linting rules to detect common non-deterministic patterns
+- Testing utilities to validate deterministic behavior
+- Runtime warnings for detected non-deterministic operations (where feasible)
+
+## 15. API Client Requirements
+
+### 15.1 Required APIs
+
+The SDK MUST use the following Lambda APIs:
+
+- `CheckpointDurableExecution`: Persist operation state
+- `GetDurableExecutionState`: Retrieve execution history
+- `SendDurableExecutionCallbackSuccess`: Complete a callback successfully
+- `SendDurableExecutionCallbackFailure`: Complete a callback with error
+- `SendDurableExecutionCallbackHeartbeat`: Send callback heartbeat
+
+### 15.2 Error Handling
+
+The SDK MUST handle API errors:
+
+- `InvalidParameterValueException`: Invalid checkpoint token, operation state, etc.
+- `ResourceNotFoundException`: Execution not found
+- `ThrottlingException`: Rate limit exceeded
+- `ServiceException`: Service errors
+
+### 15.3 Retries
+
+The SDK SHOULD:
+
+1. Retry transient API errors with exponential backoff
+2. Propagate terminal errors immediately
+3. Respect Lambda's timeout to avoid infinite retries
+
+## 16. Conformance
+
+An SDK implementation conforms to this specification if it:
+
+1. **MUST** implement all operations marked as MUST
+2. **MUST** handle all required input/output formats
+3. **MUST** support checkpoint and replay correctly
+4. **MUST** implement error handling as specified
+5. **SHOULD** implement operations marked as SHOULD
+6. **MAY** implement optional extensions
+
+### 16.1 Conformance Levels
+
+- **Level 1 (Minimal)**: Core operations (STEP, WAIT, basic error handling)
+- **Level 2 (Standard)**: All required operations, serialization, logging
+- **Level 3 (Complete)**: All operations, concurrency, advanced patterns
+
+## 17. Versioning and Compatibility
+
+### 17.1 Specification Versioning
+
+This specification uses semantic versioning:
+
+- Major: Breaking changes to required behavior
+- Minor: New optional features
+- Patch: Clarifications and corrections
+
+### 17.2 Backward Compatibility
+
+SDKs SHOULD:
+
+1. Maintain backward compatibility within major versions
+2. Provide migration guides for breaking changes
+3. Support gradual adoption of new features
+
+## 18. References
+
+### 18.1 AWS Lambda Documentation
+
+- Lambda Durable Functions: https://docs.aws.amazon.com/lambda/latest/dg/durable-functions.html
+- Lambda API Reference: https://docs.aws.amazon.com/lambda/latest/api/
+- Lambda Runtime API: https://docs.aws.amazon.com/lambda/latest/dg/runtimes-api.html
+
+### 18.2 Related Standards
+
+- RFC 2119: Key words for use in RFCs to Indicate Requirement Levels
+- JSON: ECMA-404, RFC 8259
+- ISO 8601: Date and time format
+
+## Appendix A: Example State Transitions
+
+### A.1 Simple Step Execution
+
+```
+Invocation 1:
+  - Load state: []
+  - Start STEP(id="step1")
+  - Checkpoint: START step1
```

Comment:
I guess we need more guidance here around:

- AT-LEAST/MOST-ONCE
- batching/optimizations

Reply:
Yeah - that would be a good idea.